tests/data/iterators/basic_iterator_test.py

```diff
@@ -47,35 +53,133 @@ def assert_instances_are_correct(self, candidate_instances):
-        grouped_instances = [batch.instances for batch in batches]
-        assert grouped_instances == [[self.instances[0], self.instances[1]],
-                                     [self.instances[2], self.instances[3]],
-                                     [self.instances[4]]]
+        for test_instances in (self.instances, self.lazy_instances):
+            iterator = BasicIterator(batch_size=2)
+            batches = list(iterator._create_batches(test_instances, shuffle=False))
+            grouped_instances = [batch.instances for batch in batches]
+            assert grouped_instances == [[self.instances[0], self.instances[1]],
+                                         [self.instances[2], self.instances[3]],
+                                         [self.instances[4]]]
+
+    def test_few_instances_per_epoch(self):
+        # pylint: disable=protected-access
+        for test_instances in (self.instances, self.lazy_instances):
+            iterator = BasicIterator(batch_size=2, instances_per_epoch=3)
+            # First epoch: 3 instances -> [2, 1]
+            batches = list(iterator._create_batches(test_instances, shuffle=False))
+            grouped_instances = [batch.instances for batch in batches]
+            assert grouped_instances == [[self.instances[0], self.instances[1]],
+                                         [self.instances[2]]]
+            # Second epoch: 3 instances -> [2, 1]
+            batches = list(iterator._create_batches(test_instances, shuffle=False))
+            grouped_instances = [batch.instances for batch in batches]
+            assert grouped_instances == [[self.instances[3], self.instances[4]],
+                                         [self.instances[0]]]
+            # Third epoch: 3 instances -> [2, 1]
+            batches = list(iterator._create_batches(test_instances, shuffle=False))
+            grouped_instances = [batch.instances for batch in batches]
+            assert grouped_instances == [[self.instances[1], self.instances[2]],
+                                         [self.instances[3]]]
+
+    def test_many_instances_per_epoch(self):
+        # pylint: disable=protected-access
+        for test_instances in (self.instances, self.lazy_instances):
+            iterator = BasicIterator(batch_size=2, instances_per_epoch=7)
+            # First epoch: 7 instances -> [2, 2, 2, 1]
+            batches = list(iterator._create_batches(test_instances, shuffle=False))
+            grouped_instances = [batch.instances for batch in batches]
+            assert grouped_instances == [[self.instances[0], self.instances[1]],
+                                         [self.instances[2], self.instances[3]],
+                                         [self.instances[4], self.instances[0]],
+                                         [self.instances[1]]]
+
+            # Second epoch: 7 instances -> [2, 2, 2, 1]
+            batches = list(iterator._create_batches(test_instances, shuffle=False))
```

Comment:
If `shuffle=True`, will the data get shuffled on every pass through the data in this case?

Reply:
it should, I can add a test for it